<a href="https://colab.research.google.com/github/guptavaishnavi536/Document_Retrieval/blob/main/nlp_document_retrieval_for_question_answering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a id='top'></a>
# NLP - Document Retrieval for Question Answering


Question answering (QA) is a task that answers user's questions using a large collection of documents; it consists of two steps: (1) sort possible documents that contain the answer of a given question; and (2) extract the content from these documents and elaborate an answer to the user. For this project, we have explored techniques for identify the most similar document for a question, problem called _Document Retrieval_.

**Statement**: Given a question, the document retriever have to return the most likely $k$ documents that contain the answer to the question.

The dataset used for this project is the [Stanford Question Answering Dataset](https://www.kaggle.com/datasets/stanfordu/stanford-question-answering-dataset).
   


![image](https://qa.fastforwardlabs.com/images/copied_from_nb/my_icons/QAworkflow.png)
_Image from [Intro to Automated Question Answering](https://qa.fastforwardlabs.com/methods/background/2020/04/28/Intro-to-QA.html)._

In [ ]:
%%capture
!pip install wikipedia==1.4.0
!pip install scikit-learn==1.0.2
!pip install gensim==4.0.1

## What do we want to do?

We want to create a Document Retrieval, like a search tool, such as Wikipedia. Let's explore the `wikipedia` library, to see the retriever in action.

In [ ]:
import wikipedia as wiki

k = 3
question = "What are the tourist hotspots in Portugal?"

results = wiki.search(question, results=k)
print('Question:', question)
print('Pages:  ', results)

Question: What are the tourist hotspots in Portugal?
Pages:   ['Tourist attraction', 'Portugal', 'Algarve']


### Discussion

For this question, Wikipedia's Document Retrieval returned the 5 most likely pages that contain the answer to the question.

<a id="data"></a>

---
# Data Exploration

In this section, we are going to load a `json` file into a `pandas.DataFrame`. At last, elaborate our list of documents.

[Back to Top](#top)

In [ ]:
import json
import numpy as np
import pandas as pd
!pip install opendatasets

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
import opendatasets as od
import pandas

od.download(
    "https://www.kaggle.com/datasets/stanfordu/stanford-question-answering-dataset")

Please provide your Kaggle credentials to download this dataset. Learn more: http://bit.ly/kaggle-creds
Your Kaggle username: vaishnavi536
Your Kaggle Key: ··········


100%|██████████| 8.73M/8.73M [00:00<00:00, 11.3MB/s]


In [ ]:
import os

# list the available data
for dirname, _, filenames in os.walk('/content/stanford-question-answering-dataset'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/content/stanford-question-answering-dataset/train-v1.1.json
/content/stanford-question-answering-dataset/dev-v1.1.json


In [ ]:
# based on: https://www.kaggle.com/code/sanjay11100/squad-stanford-q-a-json-to-pandas-dataframe
def squad_json_to_dataframe(file_path, record_path=['data','paragraphs','qas','answers']):
    """
    input_file_path: path to the squad json file.
    record_path: path to deepest level in json file default value is
    ['data','paragraphs','qas','answers']
    """
    file = json.loads(open(file_path).read())
    # parsing different level's in the json file
    js = pd.json_normalize(file, record_path)
    m = pd.json_normalize(file, record_path[:-1])
    r = pd.json_normalize(file,record_path[:-2])
    # combining it into single dataframe
    idx = np.repeat(r['context'].values, r.qas.str.len())
    m['context'] = idx
    data = m[['id','question','context','answers']].set_index('id').reset_index()
    data['c_id'] = data['context'].factorize()[0]
    return data

In [ ]:
# loading the data
file_path = '/content/stanford-question-answering-dataset/train-v1.1.json'
data = squad_json_to_dataframe(file_path)
data

id  \
0      5733be284776f41900661182   
1      5733be284776f4190066117f   
2      5733be284776f41900661180   
3      5733be284776f41900661181   
4      5733be284776f4190066117e   
...                         ...   
87594  5735d259012e2f140011a09d   
87595  5735d259012e2f140011a09e   
87596  5735d259012e2f140011a09f   
87597  5735d259012e2f140011a0a0   
87598  5735d259012e2f140011a0a1   

                                                question  \
0      To whom did the Virgin Mary allegedly appear i...   
1      What is in front of the Notre Dame Main Building?   
2      The Basilica of the Sacred heart at Notre Dame...   
3                      What is the Grotto at Notre Dame?   
4      What sits on top of the Main Building at Notre...   
...                                                  ...   
87594  In what US state did Kathmandu first establish...   
87595               What was Yangon previously known as?   
87596  With what Belorussian city does Kathmandu have...   
87597  In what year did Kathmandu create its initial ...   
87598                      What is KMC an initialism of?   

                                                 context  \
0      Architecturally, the school has a Catholic cha...   
1      Architecturally, the school has a Catholic cha...   
2      Architecturally, the school has a Catholic cha...   
3      Architecturally, the school has a Catholic cha...   
4      Architecturally, the school has a Catholic cha...   
...                                                  ...   
87594  Kathmandu Metropolitan City (KMC), in order to...   
87595  Kathmandu Metropolitan City (KMC), in order to...   
87596  Kathmandu Metropolitan City (KMC), in order to...   
87597  Kathmandu Metropolitan City (KMC), in order to...   
87598  Kathmandu Metropolitan City (KMC), in order to...   

                                                 answers   c_id  
0      [{'answer_start': 515, 'text': 'Saint Bernadet...      0  
1      [{'answer_start': 188, 'text': 'a copper statu...      0  
2      [{'answer_start': 279, 'text': 'the Main Build...      0  
3      [{'answer_start': 381, 'text': 'a Marian place...      0  
4      [{'answer_start': 92, 'text': 'a golden statue...      0  
...                                                  ...    ...  
87594          [{'answer_start': 229, 'text': 'Oregon'}]  18890  
87595         [{'answer_start': 414, 'text': 'Rangoon'}]  18890  
87596           [{'answer_start': 476, 'text': 'Minsk'}]  18890  
87597            [{'answer_start': 199, 'text': '1975'}]  18890  
87598  [{'answer_start': 0, 'text': 'Kathmandu Metrop...  18890  

[87599 rows x 5 columns]

In [ ]:
# how many documents do we have?
data['c_id'].unique().size

18891

## Get the Unique Documents

Let's select the unique documents in our `data`. This will be the list of documents to search for the answers.

In [ ]:
documents = data[['context', 'c_id']].drop_duplicates().reset_index(drop=True)
documents

context   c_id
0      Architecturally, the school has a Catholic cha...      0
1      As at most other universities, Notre Dame's st...      1
2      The university is the major seat of the Congre...      2
3      The College of Engineering was established in ...      3
4      All of Notre Dame's undergraduate students are...      4
...                                                  ...    ...
18886  Institute of Medicine, the central college of ...  18886
18887  Football and Cricket are the most popular spor...  18887
18888  The total length of roads in Nepal is recorded...  18888
18889  The main international airport serving Kathman...  18889
18890  Kathmandu Metropolitan City (KMC), in order to...  18890

[18891 rows x 2 columns]

<a id="document"></a>

---
# Document Retrieval

In this section, we are going to explore the techniques to retrieve documents. First, we are going to create our document `vectorizer`. We use this `vectorizer` to encode the documents and the questions into vectors. After, we can search for a question comparing with the document vectors. In the end, the algorithm will return the $k$ most similar document vectors to a question vector.


## TF-IDF

"In information retrieval, TF-IDF, short for term frequency–inverse document frequency, is a numerical statistic that is intended to reflect how important a word is to a document in a collection or corpus. It is often used as a weighting factor in searches of information retrieval, text mining, and user modeling."

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.neighbors import NearestNeighbors

# defining the TF-IDF
tfidf_configs = {
    'lowercase': True,
    'analyzer': 'word',
    'stop_words': 'english',
    'binary': True,


    'max_df': 0.9,
    'max_features': 10_000
}
# defining the number of documents to retrieve
retriever_configs = {
    'n_neighbors': 3,
    'metric': 'cosine'
}

# defining our pipeline
vectorizer = TfidfVectorizer(**tfidf_configs)
retriever = NearestNeighbors(**retriever_configs)

In [ ]:
# let's train the model to retrieve the document id 'c_id'
X = vectorizer.fit_transform(documents['context'])
retriever.fit(X, documents['c_id'])

NearestNeighbors(metric='cosine', n_neighbors=3)

Let's test the vectorizer, what information our model is using to extract the vector?

In [ ]:
def transform_text(vectorizer, text):
    '''
    Print the text and the vector[TF-IDF]
    vectorizer: sklearn.vectorizer
    text: str
    '''
    print('Text:', text)
    vector = vectorizer.transform([text])
    vector = vectorizer.inverse_transform(vector)
    print('Vect:', vector)

In [ ]:
# vectorize the question
transform_text(vectorizer, question)

Text: What are the tourist hotspots in Portugal?
Vect: [array(['tourist', 'portugal'], dtype='<U18')]


What is the most similar document to this question?

In [ ]:
# predict the most similar document
X = vectorizer.transform([question])
c_id = retriever.kneighbors(X, return_distance=False)[0][0]
print(c_id)
selected = documents.iloc[c_id]['context']
# vectorize the document
transform_text(vectorizer, selected)

987
Text: The two largest metropolitan areas have subway systems: Lisbon Metro and Metro Sul do Tejo in the Lisbon Metropolitan Area and Porto Metro in the Porto Metropolitan Area, each with more than 35 km (22 mi) of lines. In Portugal, Lisbon tram services have been supplied by the Companhia de Carris de Ferro de Lisboa (Carris), for over a century. In Porto, a tram network, of which only a tourist line on the shores of the Douro remain, began construction on 12 September 1895 (a first for the Iberian Peninsula). All major cities and towns have their own local urban transport network, as well as taxi services.
Vect: [array(['urban', 'transport', 'towns', 'tourist', 'systems', 'supplied',
       'subway', 'shores', 'services', 'september', 'remain', 'portugal',
       'porto', 'peninsula', 'network', 'mi', 'metropolitan', 'metro',
       'major', 'local', 'lisbon', 'lines', 'line', 'largest', 'km',
       'iberian', 'construction', 'cities', 'century', 'began', 'areas',
       'area',

### Evaluation

In [ ]:
%%time
# predict one document for each question
X = vectorizer.transform(data['question'])
y_test = data['c_id']
y_pred = retriever.kneighbors(X, return_distance=False)

CPU times: user 28 s, sys: 10 s, total: 38 s
Wall time: 42.4 s


In [ ]:
# top documents predicted for each question
y_pred

array([[    0,  3694, 10613],
       [    7,  1469,     2],
       [   38,  1469, 14152],
       ...,
       [18890, 18884, 18836],
       [18890,  3537, 18841],
       [    0,     1,     2]])

In [ ]:
def top_accuracy(y_true, y_pred) -> float:
    right, count = 0, 0
    for y_t in y_true:
        count += 1
        if y_t in y_pred:
            right += 1
    return right / count if count > 0 else 0

In [ ]:
acc = top_accuracy(y_test, y_pred)
print('Accuracy:', f'{acc:.4f}')
print('Quantity:', int(acc*len(y_pred)), 'from', len(y_pred))

Accuracy: 0.9892
Quantity: 86650 from 87599



1. This is a difficult problem, because we have multiples documents (in this notebook, ~19k documents) and the answer can be in one or more documents. Thus, the retriever usually returns $k$ documents, because it is not complete/fair return only one document.
2. We reach a high accuracy with top-3 (98.92%); in top-1 a low accuray (43.22%) becase we have a lot of documents, and some are pretty similar.
3. TF-IDF has some problems: (1) this algorithm is only able to compute similarity between questions and documents that present the same words, so it can not capture synonyms; and (2) cannot understand the question context or the meaning of the words.

## Word2Vec / Embedding

"Word2vec is a technique for natural language processing published in 2013. The word2vec algorithm uses a neural network model to learn word associations from a large corpus of text. Once trained, such a model can detect synonymous words or suggest additional words for a partial sentence."

In [ ]:
from gensim.parsing.preprocessing import preprocess_string

# create a corpus of tokens
corpus = documents['context'].tolist()
corpus = [preprocess_string(t) for t in corpus]

/usr/local/lib/python3.9/dist-packages/gensim/similarities/__init__.py:15: UserWarning: The gensim.similarities.levenshtein submodule is disabled, because the optional Levenshtein package <https://pypi.org/project/python-Levenshtein/> is unavailable. Install Levenhstein (e.g. `pip install python-Levenshtein`) to suppress this warning.
  warnings.warn(msg)


In [ ]:
from gensim.models import Word2Vec
import gensim.downloader

# we can download a pretrained Word2Vec
# - or we can train your own model

# download a model
# vectorizer = gensim.downloader.load('word2vec-ruscorpora-300')

# train your own model
vectorizer = Word2Vec(sentences=corpus, vector_size=300, window=5, min_count=1, workers=4).wv

In [ ]:
# similar words to 'tourist'
vectorizer.most_similar('tourist', topn=5)

[('destin', 0.9698262214660645),
 ('visitor', 0.9390003085136414),
 ('world’', 0.9342949986457825),
 ('melbourn', 0.9262661337852478),
 ('philadelphia', 0.92551189661026)]

In [ ]:
def transform_text2(vectorizer, text, verbose=False):
    '''
    Transform the text in a vector[Word2Vec]
    vectorizer: sklearn.vectorizer
    text: str
    '''
    tokens = preprocess_string(text)
    words = [vectorizer[w] for w in tokens if w in vectorizer]
    if verbose:
        print('Text:', text)
        print('Vector:', [w for w in tokens if w in vectorizer])
    elif len(words):
        return np.mean(words, axis=0)
    else:
        return np.zeros((300), dtype=np.float32)

In [ ]:
# just testing our Word2Vec
transform_text2(vectorizer, question, verbose=True)

Text: What are the tourist hotspots in Portugal?
Vector: ['tourist', 'hotspot', 'portug']


In [ ]:
# let's train the model to retrieve the document id 'c_id'
retriever = NearestNeighbors(**retriever_configs)



In [ ]:
# vectorizer the documents, fit the retriever
X = documents['context'].apply(lambda x: transform_text2(vectorizer, x)).tolist()
retriever.fit(X, documents['c_id'])

NearestNeighbors(metric='cosine', n_neighbors=3)

### Evaluation

In [ ]:
%%time
# vectorizer the questions
X = data['question'].apply(lambda x: transform_text2(vectorizer, x)).tolist()

# predict one document for each question
y_test = data['c_id']
y_pred = retriever.kneighbors(X, return_distance=False)

CPU times: user 55.3 s, sys: 4.55 s, total: 59.8 s
Wall time: 48.8 s


In [ ]:
# top documents predicted for each question
y_pred

array([[ 2637, 18125, 17575],
       [ 2424,  7290,  4816],
       [10619,  7019,     0],
       ...,
       [ 7933,  9491,   686],
       [ 2844,  1623,  1624],
       [13316, 15409, 11381]])

In [ ]:
acc = top_accuracy(y_test, y_pred)
print('Accuracy:', f'{acc:.4f}')
print('Quantity:', int(acc*len(y_pred)), 'from', len(y_pred))

Accuracy: 0.9711
Quantity: 85068 from 87599



1. We also reach a good accuracy (97.12%) in top-3; and a really low accuray (3.12%) in top-1. Thus, the TF-IDF was better.
2. Maybe, the vectorizer didn't receive enough data to be trained. Thus, pretrained models, like 'word2vec-google-news-300' could be used.



---
# Conclusion

TF-IDF reached a great top-3 accuracy (98.92%) as compared to Word2Vec for this dataset, and it can increases returning more documents. We also have other algorithms to work with Document Retriveal, such as [BM25](https://pypi.org/project/rank-bm25/) and [DPR](https://aclanthology.org/2020.emnlp-main.550/).

[Back to Top](#top)

## Reference

1. [Intro to Automated Question Answering](https://qa.fastforwardlabs.com/methods/background/2020/04/28/Intro-to-QA.html)
2. [Building a QA System with BERT on Wikipedia](https://qa.fastforwardlabs.com/pytorch/hugging%20face/wikipedia/bert/transformers/2020/05/19/Getting_Started_with_QA.html)
3. [Evaluating QA: Metrics, Predictions, and the Null Response](https://qa.fastforwardlabs.com/no%20answer/null%20threshold/bert/distilbert/exact%20match/f1/robust%20predictions/2020/06/09/Evaluating_BERT_on_SQuAD.html)
4. [Dense Passage Retrieval for Open-Domain Question Answering](https://aclanthology.org/2020.emnlp-main.550/)